SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs

```diff
@@ -190,12 +191,12 @@ public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source,
     /// A sequence of elements containing the <paramref name="source"/> and all subsequent elements 
     /// until the <paramref name="nextElementSelector"/> returns <see langword="null" />.
     /// </returns>
-    public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source, Func<TSource, TSource> nextElementSelector)
+    public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source, Func<TSource, TSource?> nextElementSelector)
         where TSource : class
     {
       ArgumentUtility.CheckNotNull ("nextElementSelector", nextElementSelector);
 
-      return CreateSequence (source, nextElementSelector, e => e != null);
+      return CreateSequence (source, nextElementSelector!, e => e != null)!;

```

Comment:
Otherwise `Func<T, T?>` wouldn't match Func<T, T> in the overload. This is necessary because we can't specify the nullability of T in Func<T, T> when T is unbounded.

Reply:
Nope, still don't understand. nextElementSelector is a delegate. How does the nullablity of the Delegate's type parameters figure into the nullablity of the delegate instance itself? No, didn't check it out to try, so I don't know the error message etc.

Reply:
Okay, and second question: why do we have a NFO on the return value of the nested CreateSequence call? That shouldn't be nullable anyway, so why the NFO?

Reply:
the `!` operator basically erases Nullability information of a type so `Func<T, T?>` gets converted to `Func<T, T>`.
The nextElementSelector here has to return null sometime otherwise the sequence won't be terminated. The `!` here is so that we can tell the compiler that we are smarter and check for the null values (in the predicate) that would otherwise throw NREs.

Reply:
What the? By supplying the NFO on a variable, we change the way the generic type parameters that are part of the type itself behave, isntead of just how the type as a whole behaves? This is completely...I mean, how? Why? Okay, I'm pretty sure that as a good reason for doing this, but why can't I do some sort of casting or something? THis is just not intuitive :/